File: 3639-zero-array-transformation-i/3639-zero-array-transformation-i.cpp
class Solution {
public:
    bool isZeroArray(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size();
        int m = queries.size();

        vector<int> arr(n+1, 0);
        for(int i=0; i<m; i++){
            auto q = queries[i];
            arr[q[0]]++;
            arr[q[1]+1]--;
        }

        for(int i=1; i<n; i++){
            arr[i]+=arr[i-1];
        }
        for(int i=0; i<n; i++){
            nums[i]-=arr[i];
            if(nums[i]>0){
                return false;
            } 
        }
        return true;
    }
};

File: 3639-zero-array-transformation-i/3639-zero-array-transformation-i.rs
impl Solution {
    pub fn is_zero_array(nums: Vec<i32>, queries: Vec<Vec<i32>>) -> bool {
        let n = nums.len();
        let m = queries.len();

        let mut arr = vec![0;n+2];

        for i in 0..m {
            arr[queries[i][0] as usize] += 1;
            arr[(queries[i][1] + 1) as usize] -= 1;

        }
        for i in 1..n {
            arr[i] += arr[i-1];
        }

        for i in 0..n {
            if (nums[i] - arr[i]) > 0 {
                return false;
            }
        }
        true
    }
}